lapce-core/src/language.rs

```diff
@@ -4,179 +4,388 @@ use tree_sitter::{Parser, TreeCursor};
+//        // ...
+//        #[cfg(feature = "lang-foo")]
+//        Setting{
+//            id: LapceLanguage::Foo,
+//            language: tree_sitter_foo::language,
+//            highlight: tree_sitter_foo::HIGHLIGHT_QUERY,
+//            comment: "//",
+//            indent: "    ",
+//            code_lens: (&[/* ... */], &[/* ... */]),
+//        },
+//    ];
+//
+// 5. Add a new match arm in `LapceLanguage::from_path`, guard the new arm with
+//    the new feature.
+//
+//    Some(match extension.as_str() {
+//        // ...
+//        #[cfg(feature = "lang-foo")]
+//        "foo" => LapceLanguage::Foo,
+//        _ => return None,
+//    })
+//
+// 6. Add a new feature, say "lang-foo", to the lapce-ui crate (see
+//    lapce-ui/Cargo.toml).
+//
+//    [features]
+//    # ...
+//    lang-foo = "lapce-core/lang-foo"
+//
+
+// Use these lists when a language does not have specific settings for "code
+// lens".
+#[allow(dead_code)]
 const DEFAULT_CODE_LENS_LIST: &[&str] = &["source_file"];
+#[allow(dead_code)]
 const DEFAULT_CODE_LENS_IGNORE_LIST: &[&str] = &["source_file"];
-const RUST_CODE_LENS_LIST: &[&str] =
-    &["source_file", "impl_item", "trait_item", "declaration_list"];
-const RUST_CODE_LENS_IGNORE_LIST: &[&str] =
-    &["source_file", "use_declaration", "line_comment"];
-const GO_CODE_LENS_LIST: &[&str] = &[
-    "source_file",
-    "type_declaration",
-    "type_spec",
-    "interface_type",
-    "method_spec_list",
-];
-const GO_CODE_LENS_IGNORE_LIST: &[&str] =
-    &["source_file", "comment", "line_comment"];
-const PYTHON_CODE_LENS_LIST: &[&str] = &[
-    "source_file",
-    "module",
-    "class_definition",
-    "class",
-    "identifier",
-    "decorated_definition",
-    "block",
-];
-const PYTHON_CODE_LENS_IGNORE_LIST: &[&str] =
-    &["source_file", "import_statement", "import_from_statement"];
-const JAVASCRIPT_CODE_LENS_LIST: &[&str] = &["source_file", "program"];
-const JAVASCRIPT_CODE_LENS_IGNORE_LIST: &[&str] = &["source_file"];
 
-#[derive(Eq, PartialEq, Hash, Copy, Clone, Debug)]
+struct Setting {
+    /// An extra check to make sure that the array elements are in the correct
+    /// order.  If this id does not match the enum value, a panic will happen
+    /// with a debug assertion message.
+    id: LapceLanguage,
+    /// This is the factory function defined in the tree-sitter crate that
+    /// creates the language parser.  For most languages, it is
+    /// `tree_sitter_$crate::language`.
+    language: fn() -> tree_sitter::Language,
+    /// For most languages, it is `tree_sitter_$crate::HIGHLIGHT_QUERY`.
+    highlight: &'static str,
+    /// The comment token.  "#" for python, "//" for rust for example.
+    comment: &'static str,
+    /// The indent unit.  "\t" for python, "    " for rust, for example.
+    indent: &'static str,
+    /// If unsure, use `DEFAULT_CODE_LENS_LIST` and
+    /// `DEFAULT_CODE_LENS_IGNORE_LIST`.
+    code_lens: (&'static [&'static str], &'static [&'static str]),
```

Comment:
Can we have two fields here instead of a tuple?

Reply:
> Can we have two fields here instead of a tuple?

Sure.

Reply:
I was trying to add some comments to explain the tuple. But I don't know what to say about the two lists without knowing tree-sitter. I will leave a "TODO" in the comment for someone else to address it.